### PR TITLE
Add remote write feature for Shoot Prometheus

### PR DESCRIPTION
**Pull request by dergeberl**
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:
This PR enables the possibility to set an remote write endpoint for Shoot Prometheus to collect all metrics on a central place (for example cortex).

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
I`m not sure if the place for the secret handling is the right one. It looks like a good place for it but it is hard to follow the way of a secret.

**Release note**:

```feature operator
Allow configuration of remote write endpoint for Shoot Prometheus to collect all Shoot metrics on a central instance
```

### CODE REVIEW

**Comment by gardener-robot-ci-1** [comment]
Thank you @dergeberl for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: docs/monitoring/README.md [hunk 1]
```diff
@@ -49,3 +49,17 @@ Deployed in the shoot control plane namespace. Important scrape targets:
...
+
+## Collect all Shoot Prometheus with remote write
+
+An optional collection of all Shoot Prometheus metrics to a central prometheus (or cortex) instance is possible with the `monitoring` setting in `cloudprofile`:
+```
+  monitoring:
+    externalLabels: # add additional labels to metrics to identify it on the central instance
+      additional: label
+    remoteWriteURL: https://remoteWriteUrl # remote write URL
+    remoteWriteKeep: # metrics that should be forwarded to the external write endpint. Im empty all metrics get forwarded
```
Comment (wyb1) [inline docs/monitoring/README.md]:
```suggestion
    remoteWriteKeep: # metrics that should be forwarded to the external write endpoint. If empty all metrics get forwarded
```

File: example/30-cloudprofile.yaml [hunk 2]
```diff
@@ -47,6 +47,12 @@ spec:
   #   size: 20Gi     # optional, either size or minSize must be configured
   #   minSize: 10Gi  # optional, either size or minSize must be configured
     usable: true
+# monitoring:
+#   externalLabels: # add additional labels to metrics to identify it on the central instance
+#     additional: label
+#   remoteWriteURL: https://remoteWriteUrl # remote write URL
+#   remoteWriteKeep: # metrics that should be forwarded to the external write endpint. Im empty all metrics get forwarded
```
Comment (wyb1) [inline example/30-cloudprofile.yaml]:
```suggestion
#   remoteWriteKeep: # metrics that should be forwarded to the external write endpoint. If empty all metrics get forwarded
```

**Review by wyb1 — COMMENTED** [review]
Very nice! 🎉  Thanks for the the PR! Overall it looks good, but I haven't tried it out yet. Will try it out later today

File: charts/seed-monitoring/charts/core/charts/prometheus/templates/config.yaml [hunk 3]
```diff
@@ -26,6 +26,29 @@ data:
...
+    {{- end }}
+    {{- if .Values.remoteWrite.keep }}
+      write_relabel_configs:
+{{ include "prometheus.keep-metrics.metric-relabel-config" .Values.remoteWrite.keep | indent 6 }}
+    {{- end }}
+      queue_config:
+          max_shards: 100
+          batch_send_deadline: 20s
+          min_backoff: 500ms
+          max_backoff: 60s
```
Comment (wyb1) [inline charts/seed-monitoring/charts/core/charts/prometheus/templates/config.yaml]: Can we make this configurable? I am not sure what the best way to do this is. Perhaps simply a `queue_config` key in the `cloudprofile` that can contain a map since gardener should not validate all of the prometheus config (@rfranzke any thoughts?). If it is empty this should be the default.

File: docs/monitoring/README.md [hunk 4]
```diff
@@ -49,3 +49,17 @@ Deployed in the shoot control plane namespace. Important scrape targets:
...
+## Collect all Shoot Prometheus with remote write
+
+An optional collection of all Shoot Prometheus metrics to a central prometheus (or cortex) instance is possible with the `monitoring` setting in `cloudprofile`:
+```
+  monitoring:
+    externalLabels: # add additional labels to metrics to identify it on the central instance
+      additional: label
+    remoteWriteURL: https://remoteWriteUrl # remote write URL
+    remoteWriteKeep: # metrics that should be forwarded to the external write endpoint. If empty all metrics get forwarded
+    - kube_pod_container_info
```
Comment (wyb1) [inline docs/monitoring/README.md]: Please add a top level shoot key. We will probably extend this for the prometheus instances responsible for the seed. The names of the variables `global-remote-write-monitoring` should be adjusted as well to make it clear that it belongs to the shoot prometheus.
 
```suggestion
    shoot:
      externalLabels: # add additional labels to metrics to identify it on the central instance
        additional: label
      remoteWriteURL: https://remoteWriteUrl # remote write URL
      remoteWriteKeep: # metrics that should be forwarded to the external write endpoint. If empty all metrics get forwarded
      - kube_pod_container_info
```

File: example/10-secret-remote-write.yaml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+metadata:
+  name: monitoring-remote-write-credentials
+  namespace: garden
+  labels:
+    gardener.cloud/role: global-remote-write-monitoring
+type: Opaque
+data:
+   # Basic Auth
+  username: base64(admin)
+  password: base64(password)
```
Comment (wyb1) [inline example/10-secret-remote-write.yaml]: newline

File: charts/seed-monitoring/charts/core/charts/prometheus/values.yaml [hunk 6]
```diff
@@ -307,3 +307,17 @@ resources:
...
+#    username: user
+#    password: password
+#  keep:
+#  - metric1
+#  - metric2
+#  - metric3
+#  - metric4
+
+#externalLabels:
+#  env: test
```
Comment (wyb1) [inline charts/seed-monitoring/charts/core/charts/prometheus/values.yaml]: newline

File: docs/monitoring/README.md [hunk 7]
```diff
@@ -49,3 +49,17 @@ Deployed in the shoot control plane namespace. Important scrape targets:
...
 
 **Purpose**: Monitor all relevant components belonging to a shoot cluster managed by Gardener. Shoot owners can view the metrics in Grafana dashboards and receive [alerts](user_alerts.md) based on these metrics. Gardener operators will receive a different set of [alerts](operator_alerts.md). For alerting internals refer to [this](alerting.md) document.
+
+## Collect all Shoot Prometheus with remote write
+
+An optional collection of all Shoot Prometheus metrics to a central prometheus (or cortex) instance is possible with the `monitoring` setting in `cloudprofile`:
+```
+  monitoring:
+    externalLabels: # add additional labels to metrics to identify it on the central instance
+      additional: label
```
Comment (wyb1) [inline docs/monitoring/README.md:69]: This will be the same for every prometheus in an entire landscape if I am not mistaken. Do you want to add a "landscape" label and send the data to a central cortex/prometheus from multiple landscapes? Just trying to understand the use case.

**Review by wyb1 — CHANGES_REQUESTED** [review]
Tried it out and everything worked as expected :) 
Just have some comments regarding the API.

File: docs/monitoring/README.md (hunk 7 above)
Comment (dergeberl) [reply docs/monitoring/README.md:69]: Yes the use case is to identify a landscape for example a test and production landscape.

File: charts/seed-monitoring/charts/core/charts/prometheus/templates/config.yaml (hunk 3 above)
Comment (dergeberl) [reply charts/seed-monitoring/charts/core/charts/prometheus/templates/config.yaml]: I added queue_config to the cloudprofile as multiline string, because of multiple types (string, int and bool) so `map[string]string` will not work. We did not find a solution than can handle multiple types and works with protobuf.

File: docs/monitoring/README.md (hunk 4 above)
Comment (dergeberl) [reply docs/monitoring/README.md]: I added shoot and restructured `remoteWrite` in a own struct. I excluded `externalLabels` because it can be used also in for example federation.

**Review by wyb1 — DISMISSED** [review]
/lgtm

**Comment by rfranzke** [comment]
Sorry for joining late here, but why is this configuration part of of the `CloudProfile` resource? This one typically describes cloud provider specific settings only, and this monitoring stuff isn't of this kind, is it? Wouldn't the [gardenlet component configuration](https://github.com/gardener/gardener/blob/master/example/20-componentconfig-gardenlet.yaml) be a better fit? For example, there is already similar configuration for the logging stuff: https://github.com/gardener/gardener/blob/3262c48d5c76cc53b78b4af02d107656d67a4492/example/20-componentconfig-gardenlet.yaml#L110-L120

**Comment by dergeberl** [comment]
@rfranzke yes, that sounds good. I moved the config from Cloudprofile to GardenletConfiguration.

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 8]
```diff
@@ -507,6 +510,35 @@ type LoadBalancerServiceConfig struct {
...
+	// Shoot is optional and contains settings for the shoot monitoring stack.
+	// +optional
+	Shoot *ShootMonitoringConfig `json:"shoot,omitempty"`
+}
+
+// ShootMonitoringConfig contains settings for the shoot monitoring stack.
+type ShootMonitoringConfig struct {
+	// RemoteWrite is optional and contains remote write setting.
+	// +optional
+	RemoteWrite RemoteWriteMonitoringConfig `json:"remoteWrite,omitempty"`
```
Comment (rfranzke) [inline pkg/gardenlet/apis/config/v1alpha1/types.go]: If this is optional then it should be a pointer type, otherwise (if it's required) please remove `omitempty` and the `// +optional` tag

File: pkg/gardenlet/apis/config/v1alpha1/types.go [hunk 9]
```diff
@@ -507,6 +510,35 @@ type LoadBalancerServiceConfig struct {
...
+// RemoteWriteMonitoringConfig contains settings for the remote write setting for monitoring stack.
+type RemoteWriteMonitoringConfig struct {
+	// URL contains an Url for remote write setting in prometheus.
+	URL string `json:"url"`
+	// Keep contains a list of metrics that will be remote written
+	// +optional
+	Keep []string `json:"keep,omitempty"`
+	// QueueConfig contains the queue_config for prometheus remote write.
+	// +optional
+	QueueConfig string `json:"queueConfig,omitempty"`
```
Comment (rfranzke) [inline pkg/gardenlet/apis/config/v1alpha1/types.go]: same here

**Review by rfranzke — COMMENTED** [review]
Looks good, just two minor nits

**Review by rfranzke — DISMISSED** [review]
/lgtm
/reviewed ok-to-test

**Comment by rfranzke** [comment]
@dergeberl Please check the failing CI:

> Executing gofmt/goimports
Unformatted files detected:
pkg/operation/botanist/monitoring.go

**Review by rfranzke — APPROVED** [review]
